spread/openstack.go

```diff
@@ -0,0 +1,631 @@
+	}
+	return fmt.Sprintf("%s (%s)", s.system, s.d.Name)
+}
+
+func (s *openstackServer) Label() string {
+	return s.d.Name
+}
+
+func (s *openstackServer) Provider() Provider {
+	return s.p
+}
+
+func (s *openstackServer) Address() string {
+	return s.address
+}
+
+func (s *openstackServer) System() *System {
+	return s.system
+}
+
+func (s *openstackServer) ReuseData() interface{} {
+	return &s.d
+}
+
+const (
+	openstackStaging      = "STAGING"
+	openstackProvisioning = "PROVISIONING"
+	openstackRunning      = "RUNNING"
+	openstackStopping     = "STOPPING"
+	openstackStopped      = "STOPPED"
+	openstackSuspending   = "SUSPENDING"
+	openstackTerminating  = "TERMINATED"
+
+	openstackPending = "PENDING"
+	openstackDone    = "DONE"
+)
+
+func (p *openstackProvider) Backend() *Backend {
+	return p.backend
+}
+
+func (p *openstackProvider) Reuse(ctx context.Context, rsystem *ReuseSystem, system *System) (Server, error) {
+	s := &openstackServer{
+		p:       p,
+		address: rsystem.Address,
+		system:  system,
+	}
+	err := rsystem.UnmarshalData(&s.d)
+	if err != nil {
+		return nil, fmt.Errorf("cannot unmarshal openstack reuse data: %v", err)
+	}
+	return s, nil
+}
+
+func (p *openstackProvider) Allocate(ctx context.Context, system *System) (Server, error) {
+	if err := p.checkKey(); err != nil {
+		return nil, err
+	}
+
+	s, err := p.createMachine(ctx, system)
+	if err != nil {
+		return nil, err
+	}
+
+	printf("Allocated %s.", s)
+	return s, nil
+}
+
+func (s *openstackServer) Discard(ctx context.Context) error {
+	return s.p.removeMachine(ctx, s)
+}
+
+const openstackCloudInitScript = `
+#cloud-config
+runcmd:
+  - echo root:%s | chpasswd
+  - sed -i 's/^\s*#\?\s*\(PermitRootLogin\|PasswordAuthentication\)\>.*/\1 yes/' /etc/ssh/sshd_config
+  - test -d /etc/ssh/sshd_config.d && echo 'PermitRootLogin=yes' > /etc/ssh/sshd_config.d/00-spread.conf
+  - test -d /etc/ssh/sshd_config.d && echo 'PasswordAuthentication=yes' >> /etc/ssh/sshd_config.d/00-spread.conf
+  - pkill -o -HUP sshd || true
+`
+const openstackReadyMarker = "MACHINE-IS-READY"
+const openstackNameLayout = "Jan021504.000000"
+const openstackDefaultFlavor = "m1.medium"
+
+type openstackImage struct {
+	Project string
+	Name    string
+	Family  string
+	Terms   []string
+}
+
+type openstackImagesCache struct {
+	mu     sync.Mutex
+	ready  bool
+	images []openstackImage
+	err    error
+}
+
+func openstackName() string {
+	return strings.ToLower(strings.Replace(time.Now().UTC().Format(openstackNameLayout), ".", "-", 1))
+}
+
+func findErrorTitle(node *html.Node) string {
+	if node.Type == html.ElementNode && node.Data == "title" {
+		return node.FirstChild.Data
+	}
+
+	for child := node.FirstChild; child != nil; child = child.NextSibling {
+		title := findErrorTitle(child)
+		if title != "" {
+			return title
+		}
+	}
+
+	return ""
+}
+
+// error messages returned by openstack api are html
+// which contain title and details related to the error
+func errorTitle(msg string) string {
```

Comment:
This is curious, I looked at the API in https://github.com/go-goose/goose/blob/abaee9096e3a/nova/nova.go#L130 and it would be nice too what error type you actually get here from nova. As a quick experiment `fmt.Sprintf("%v %T\n", err, err) ` would be great. I'm puzzled that this is needed.

Reply:
fmt.Sprintf("%v, err)
2023-07-17 16:26:39 authentication failed
caused by: requesting token: Unauthorised URL https://keystone.bos01.canonistack.canonical.com:5000/v3/auth/tokens
caused by: request (https://keystone.bos01.canonistack.canonical.com:5000/v3/auth/tokens) returned unexpected status: 401; error info: Failed: 401 error: The request you have made requires authentication.

fmt.Sprintf("%T, err)
2023-07-17 16:26:39 *errors.gooseError

Reply:
I could reproduce it in checkKey function